controllers/datasource_controller.go

```diff
@@ -164,19 +165,25 @@ func (r *GrafanaDatasourceReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return ctrl.Result{RequeueAfter: RequeueDelay}, err
 	}
 
-	if datasource.Spec.InstanceSelector == nil {
-		controllerLog.Info("no instance selector found for datasource, nothing to do", "name", datasource.Name, "namespace", datasource.Namespace)
-		return ctrl.Result{RequeueAfter: RequeueDelay}, err
-	}
-
 	instances, err := GetMatchingInstances(ctx, r.Client, datasource.Spec.InstanceSelector)
 	if err != nil {
 		controllerLog.Error(err, "could not find matching instance", "name", datasource.Name)
+		datasource.Status.NoMatchingInstances = true
+		err = r.Client.Status().Update(ctx, datasource)
+		if err != nil {
+			controllerLog.Error(err, "unable to update status", "name", datasource.Name)
+		}
 		return ctrl.Result{RequeueAfter: RequeueDelay}, err
 	}
 
 	if len(instances.Items) == 0 {
 		controllerLog.Info("no matching instances found for datasource", "datasource", datasource.Name, "namespace", datasource.Namespace)
+		datasource.Status.NoMatchingInstances = true
+		err = r.Client.Status().Update(ctx, datasource)
+		if err != nil {
+			controllerLog.Error(err, "unable to update status", "name", datasource.Name)
+		}
+		return ctrl.Result{RequeueAfter: RequeueDelay}, err
 	}
 
```

Comment:
Should we add `datasource.Status.NoMatchingInstances = false` somewhere around this line?